File: 912. Sort an Array/sortArrayMergeSort.py
from typing import List

class Solution:
  def merge(self, leftArr, rightArr, res):
    leftArrLen = len(leftArr)
    rightArrLen = len(rightArr)
    i, j, k = 0, 0, 0
    
    while i < leftArrLen and j < rightArrLen:
      if leftArr[i] < rightArr[j]:
        res[k] = leftArr[i]
        i += 1
      else:
        res[k] = rightArr[j]
        j += 1
      k += 1
      
    while i < leftArrLen:
      res[k] = leftArr[i]
      i += 1
      k += 1
      
    while j < rightArrLen:
      res[k] = rightArr[j]
      j += 1
      k += 1
      
  def mergeSort(self, nums):
    l = len(nums)
    
    if l < 2:
      return

    leftArr = []
    rightArr = []
    
    m = l // 2

    leftArr = nums[:m]
    rightArr = nums[m:]

    self.mergeSort(leftArr)
    self.mergeSort(rightArr)
    
    self.merge(leftArr, rightArr, nums)
    
  def sortArray(self, nums: List[int]) -> List[int]:
    self.mergeSort(nums)
    return nums


nums = [5, 2, 3, 1]
solution = Solution()
print('sortArray: ', solution.sortArray(nums))

File: 912. Sort an Array/sortArrayMergeSort.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
const sortArray = (nums) => {
  const merge = (leftArr, rightArr, res) => {
    const leftArrLen = leftArr.length;
    const rightArrLen = rightArr.length;
    let i = 0;
    let j = 0;
    let k = 0;

    while (i < leftArrLen && j < rightArrLen) {
      if (leftArr[i] < rightArr[j]) {
        res[k] = leftArr[i];
        i++;
      } else {
        res[k] = rightArr[j];
        j++;
      }
      k++;
    }

    while (i < leftArrLen) {
      res[k] = leftArr[i];
      i++;
      k++;
    }

    while (j < rightArrLen) {
      res[k] = rightArr[j];
      j++;
      k++;
    }
  }

  const mergeSort = (arr) => {
    const l = arr.length;

    if (l < 2) {
      return;
    }

    const m = Math.floor(l / 2);
    const leftArr = [];
    const rightArr = [];

    for (let i = 0; i < m; i++) {
      leftArr.push(arr[i]);
    }

    for (let j = m; j < l; j++) {
      rightArr.push(arr[j]);
    }

    mergeSort(leftArr);
    mergeSort(rightArr);

    merge(leftArr, rightArr, arr);
  }

  mergeSort(nums);
  return nums;
};


const nums = [5, 2, 3, 1];
console.log('sortArray: ', sortArray(nums));
